packages/cel/src/equals.ts

```diff
@@ -0,0 +1,222 @@
+import { equals as equalsMessage } from "@bufbuild/protobuf";
+import {
+  DurationSchema,
+  isWrapper,
+  TimestampSchema,
+} from "@bufbuild/protobuf/wkt";
+
+/**
+ * Checks for equality of two CEL values. It follows the following rules:
+ *
+ * - Primitives are checked for value equality using native equality check.
+ * - Types need not be equal for numerical values to be equal.
+ * - Wrappers are treated as scalars.
+ * - Protobuf messages are checked using protobuf equality
+ * - Lists are equal of if lengths match and each element matches the corresponding element in the other one.
+ * - Maps are equal if both the have the same set of keys and corresponding values.
+ *
+ * If the two cannot be compared, it returns false.
+ */
+export function equals(lhs: unknown, rhs: unknown): boolean {
+  // Fast path for Int, Double, Bool, and String types.
+  if (lhs === rhs) {
+    return true;
+  }
+  // Remaining scalars or scalars from Wrapper types
+  if (lhs instanceof CelUint || (isMessage(lhs) && isWrapper(lhs))) {
+    lhs = lhs.value;
+  }
+  if (rhs instanceof CelUint || (isMessage(rhs) && isWrapper(rhs))) {
+    rhs = rhs.value;
+  }
+  // Do a simple check again for uint/wrapper types.
+  if (lhs === rhs) {
+    return true;
+  }
+  // Check for scalars that are of different type and unwrapped values
+  if (
+    (typeof lhs === "number" || typeof lhs === "bigint") &&
+    (typeof rhs === "number" || typeof rhs === "bigint")
+  ) {
+    if (typeof lhs !== typeof rhs) {
+      lhs = Number(lhs);
+      rhs = Number(rhs);
+    }
+    return lhs === rhs;
+  }
+  // For values whose types must be equal for them to be equal.
+  switch (true) {
+    case lhs instanceof Uint8Array:
+      return rhs instanceof Uint8Array && equalsBytes(lhs, rhs);
+    case lhs instanceof CelList:
+      return rhs instanceof CelList && equalsList(lhs, rhs);
+    case lhs instanceof CelMap:
+      return rhs instanceof CelMap && equalsMap(lhs, rhs);
+    case lhs instanceof CelType:
+      return rhs instanceof CelType && lhs.name === rhs.name;
+    case isReflectList(lhs):
+      return isReflectList(rhs) && equalsReflectList(lhs, rhs);
+    case isReflectMap(lhs):
+      return isReflectMap(rhs) && equalsReflectMap(lhs, rhs);
+  }
+  // Proto Null
+  if (lhs instanceof ProtoNull) {
+    return (
+      rhs === null ||
+      (rhs instanceof ProtoNull && lhs.messageTypeName === rhs.messageTypeName)
+    );
+  }
+  if (rhs instanceof ProtoNull && lhs === null) {
+    return true;
+  }
+  if (lhs instanceof CelObject) {
+    lhs = lhs.value;
+  }
+  if (rhs instanceof CelObject) {
+    rhs = rhs.value;
+  }
+  // Messages
+  if (isMessage(lhs)) {
+    if (isWrapper(lhs)) {
+      lhs = lhs.value;
+    } else {
+      lhs = reflect(getSchema(lhs.$typeName), lhs);
+    }
+  }
+  if (isMessage(rhs)) {
+    if (isWrapper(rhs)) {
+      rhs = rhs.value;
+    } else {
+      rhs = reflect(getSchema(rhs.$typeName), rhs);
+    }
+  }
```

Comment:
Ideally we can remove these where we only expect reflect or scalar values for primitives.